### PR TITLE
Add unit testing to the project

### PR DESCRIPTION
**Pull request by anand-kamble**
## Execution
I have made a test class that can be used for testing. 
You can try running the program by switching to this branch.
```bash 
git fetch --all && git checkout copa-14-add-unit-testing-to-the-project
```
Update the poetry environment packages:
```bash
poetry install
```
>There is a additional dev dependency of `pytest`  

And then running the tests by:
```bash
poetry run ./run.sh --test
```

Output like below is expected:
image

## Details
This PR includes the new Test class which can be used to write unit tests as needed. 
I have designed the methods such that it's easier to read the test and new tests can be written quickly.

### Example:
Creating a new test 
```python
my_test = Test("My unit test", mode="soft")
```
* 1st argument is the title given to this test.
* `soft` mode means testing won't stop even if one test case fails. Another option is `hard` which will result in program exiting with code 1 if a test case fails.

Let us use some numbers to test
```python
a = 1
b = 2
```
Now, if we have to check if the value of variable a is 1, we would write like this:
```python
my_test.expect(a).to_be(1)
```
These tests are also chainable and can be called like:
```python
my_test.expect(b).to_be(2).to_be_of_type(int)
```
This will make it easier for us to write tests in the future.
You can check the implementation in the file `tests/main.py`

### CODE REVIEW

File: Readme.md [hunk 1]
```diff
@@ -21,14 +21,9 @@ You'll need poetry to create the python virtual environment to run this project.
 
 Once you have installed poetry, follow this steps to generate the filtered data.
```
Comment (erlebach) [inline Readme.md]: Follow "these" steps.

File: tests/test.py [hunk 2]
```diff
@@ -11,6 +11,20 @@ class Test:
     """
     A class for writing and running tests with a fluent API.
```
Comment (erlebach) [inline tests/test.py]: What do you mean by "fluent" API?  You mean "easy to use"?

File: tests/test.py [hunk 3]
```diff
@@ -11,6 +11,20 @@ class Test:
...
 
+    Attributes:
+        title (str): The title of the test.
+        mode (Literal["soft", "hard"]): The mode of the test. If set to "hard", the test will exit upon the first failure.
+        export_file_name (str): The name of the file where the test results will be exported.
+        result (str): The result of the test, formatted as a Markdown string.
```
Comment (erlebach) [inline tests/test.py]: result (str): Test results, formatted as a Markdown string.

File: tests/test.py [hunk 4]
```diff
@@ -11,6 +11,20 @@ class Test:
...
+        test_obj: The object that is currently being tested.
+
+    Methods:
+        expect(obj): Set the object to be tested.
+        to_be(expected, msg: str | None = None): Assert that the test object is equal to the expected value.
+        to_be_of_type(obj_type, msg: str | None = None): Assert that the test object is an instance of the expected type.
+        to_be_approximately(expected, precision, msg: str | None = None): Assert that the test object is approximately equal to the expected value within a given precision.
```
Comment (erlebach) [inline tests/test.py]: "Why not "is_approximately" rather "to_be_approximately". Short declarative names are better. Same comment for the other functions.

**Review by erlebach — COMMENTED** [review]
I have not tried the code yet.

File: Readme.md (hunk 1 above)
Comment (erlebach) [reply Readme.md]: When I ran `poetry install`, I got the message: 
```
Warning: The current project could not be installed: No file/folder found for package copa-data-filtering
```
This message went away once I created the folder 'copa-data-filtering'. Are you using this folder?

File: tests/test.py (hunk 4 above)
Comment (anand-kamble) [reply tests/test.py]: Will update that.

File: tests/test.py (hunk 2 above)
Comment (anand-kamble) [reply tests/test.py]: Yes. I meant "easy to use".

**Review by erlebach — COMMENTED** [review]
My `pyproject.toml` file did not contain `test` in the `dev` section. That would imply that when I switched to your branch, that the `pyproject.toml` did not download. 

I removed my lock file and .venv. I reran 'poetry install' and got the message: 
```
Installing the current project: copa-data-filtering (1.0.4)
Warning: The current project could not be installed: No file/folder found for package copa-data-filtering
If you do not want to install the current project use --no-root.
If you want to use Poetry only for dependency management but not for packaging, you can disable package mode by setting package-mode = false in your pyproject.toml file.
In a future version of Poetry this warning will become an error!
```

I read, contrary to what I thought, that the pyproject.lock file should be included in the repository. So include the lock file and let me know. I'll then I'll take another look.

File: tests/test.py (hunk 4 above)
Comment (anand-kamble) [reply tests/test.py]: Completed in commit https://github.com/anand-kamble/data-filtering/pull/9/commits/2a83b5701d708ed985be33b130892e05e438acea

**Comment by anand-kamble** [comment]
> My `pyproject.toml` file did not contain `test` in the `dev` section. That would imply that when I switched to your branch, that the `pyproject.toml` did not download.
> 
> I removed my lock file and .venv. I reran 'poetry install' and got the message:
> 
> ```
> Installing the current project: copa-data-filtering (1.0.4)
> Warning: The current project could not be installed: No file/folder found for package copa-data-filtering
> If you do not want to install the current project use --no-root.
> If you want to use Poetry only for dependency management but not for packaging, you can disable package mode by setting package-mode = false in your pyproject.toml file.
> In a future version of Poetry this warning will become an error!
> ```
> 
> I read, contrary to what I thought, that the pyproject.lock file should be included in the repository. So include the lock file and let me know. I'll then I'll take another look.

I am not using any folder named `copa-data-filtering`. 
And have added my lock file in this branch at https://github.com/anand-kamble/data-filtering/pull/9/commits/7427a8e15a163d35941700e955cdeec15bc5a82f. 

Please pull the latest code by:
```bash
git pull origin copa-14-add-unit-testing-to-the-project
```

File: tests/test.py (hunk 3 above)
Comment (anand-kamble) [reply tests/test.py]: Fixed in https://github.com/anand-kamble/data-filtering/pull/9/commits/d639c735d3a60db7d51c2bfc9c9cc1b74622c48c

File: tests/test.py (hunk 2 above)
Comment (anand-kamble) [reply tests/test.py]: Updated in https://github.com/anand-kamble/data-filtering/pull/9/commits/d639c735d3a60db7d51c2bfc9c9cc1b74622c48c

**Comment by anand-kamble** [comment]
I had to rebase this branch, which is why there are some duplicate commits.

**Comment by erlebach** [comment]
Yep. We should merge everything into main as things are finalized.

____________________________
Gordon Erlebacher
Director, Interdisciplinary Program in Data Science
Florida State University

**Comment by anand-kamble** [comment]
Done. Please check the latest version of the main​ branch.
________________________________

**Comment by erlebach** [comment]
Add an "INSTALL" file.

How to write a good INSTALL file:

https://chatgpt.com/share/7dcddae0-f7d8-4e30-bd5b-058ae382609e

**Comment by erlebach** [comment]
I ran a program written by GPT-4o on ATA_21-merged and got 5 figures:

https://chatgpt.com/share/f644fc4a-08ce-4c15-b91f-9a13b0ce65f0

here are your plots. I am sure you can do much better.






[image: 2024-06-04_21-05-32.png][image: 2024-06-04_21-05-22.png][image:
2024-06-04_21-05-16.png][image: 2024-06-04_21-05-11.png][image:
2024-06-04_21-05-02.png]

**Comment by anand-kamble** [comment]
I will update the code accordingly.

Anand.
________________________________

**Comment by anand-kamble** [comment]
Good Morning Prof.,

I have added the INSTALL.md<https://github.com/anand-kamble/data-filtering/blob/main/INSTALL.md> file to the repository. I am looking at what is the best way to have the notebooks in the repository.
According to Perplexity, https://www.perplexity.ai/search/I-have-a-FVppDuiiRXm40LRqUZDN8Q
It just says to create a new folder named "notebooks/" ​

________________________________

**Comment by erlebach** [comment]
Since you are using poetry, look at best oractu es for your.

Cline langchain and llamas and apply tree with -d option to better
understand how the experts do it.

____________________________
Gordon Erlebacher
Director, Interdisciplinary Program in Data Science
Florida State University

**Comment by erlebach** [comment]
Good work on INSTALL.

Some comments:

1) the repo is private, so git clone will not work. But leave it. Add
comment to request access to the git repo.

2) under src, you could have a folder for all the subprojects occurring in
our project:

- ATA_exploration
- ATA generation
etc.

Later, we'll consolidate.

**Comment by anand-kamble** [comment]
[like]  Anand Kamble reacted to your message:
________________________________

**Pull request merged**